### PR TITLE
Saga analyzers

### PR DESCRIPTION
**Pull request by DavidBoike**

### CODE REVIEW

File: src/NServiceBus.Core.Analyzer/Sagas/SagaDiagnostics.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            title: "Saga mapping expressions can be simplified with newer syntax",
+            messageFormat: "Message: The saga mapping contains multiple .ToSaga(…) expressions which can be simplified using mapper.MapSaga(…).ToMessage<T>(…) syntax.",
+            category: DiagnosticCategory,
+            defaultSeverity: DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        internal static readonly DiagnosticDescriptor MultipleCorrelationIdValues = new DiagnosticDescriptor(
+            id: MultipleCorrelationIdValuesId,
+            title: "Saga can only define a single correlation property on the saga data",
+            messageFormat: "The saga can only map the correlation id to one property on the saga data class.",
```
Comment (cquirosj) [inline src/NServiceBus.Core.Analyzer/Sagas/SagaDiagnostics.cs:44]:
```suggestion
            messageFormat: "The saga can only map the correlation ID to one property on the saga data class.",
```

File: src/NServiceBus.Core.Analyzer.Tests/Sagas/RewriteConfigureHowToFindSagaFixerTests.cs [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        //    }
+        //    public class Msg2 : ICommand
+        //    {
+        //        public string CorrId { get; set; }
+        //    }
+        //}".Replace("    ", "\t");
+
+        //            return Assert(original, expected);
+        //        }
+
```
Comment (cquirosj) [inline src/NServiceBus.Core.Analyzer.Tests/Sagas/RewriteConfigureHowToFindSagaFixerTests.cs:301]: Do we want to leave this test code commented out here?

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        static GenericNameSyntax CreateGenericMappingMethod(string methodName, string messageType)
+        {
+            return GenericName(methodName)
+                .WithTypeArgumentList(TypeArgumentList(SingletonSeparatedList<TypeSyntax>(IdentifierName(messageType))));
+        }
+
+        static ArgumentListSyntax CreateCorrelationIdMapping(string correlationId)
+        {
+            return ArgumentList(
+                //Token(SyntaxKind.OpenParenToken),
```
Comment (cquirosj) [inline src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:178]: Leaving this comment for debugging only purposes too?

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        static async Task<Document> RewriteConfigureHowToFindSagaMethod(
+            Document document,
+            MethodDeclarationSyntax method,
+            string correlationId,
+            string mapperParamName,
+            ImmutableArray<SagaMessageMapping> mappings,
+            CancellationToken cancellationToken)
+        {
+            // default(SyntaxNode).ToString() == empty string ""
```
Comment (cquirosj) [inline src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:94]: commented out code during development

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+            for (int i = 0; i < mappings.Length; i++)
+            {
+                var mapping = mappings[i];
+
+                mapSagaInvocation = InvocationExpression(
+                    MemberAccessExpression(
+                        SyntaxKind.SimpleMemberAccessExpression,
+                        mapSagaInvocation.WithTrailingTrivia(EndOfLine(NewLine)),
+                        Token(SyntaxKind.DotToken).WithLeadingTrivia(Indent(2)),
```
Comment (cquirosj) [inline src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:124]: Can we do something like method.GetLeadingTrivia() as an attempt to preserve the indentation from the user? Some could have spaces and other could have tabs.

Comment (cquirosj) [reply src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:124]: Never mind, looks like that is what you are determining in lines 95 and 96 to preserve the original indentation. Don't know if GetLeadingTriavia() is a "native" way to find out the same.

**Review by cquirosj — COMMENTED** [review]
Can we have the branch rebased on the latest master?

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs (hunk 4 above)
Comment (DavidBoike) [reply src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:94]: A note for the following lines. I'll change to:
```suggestion
            // Note: default(SyntaxNode).ToString() == empty string "", so you can compare the OrDefault value to string.
```

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs (hunk 5 above)
Comment (DavidBoike) [reply src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:124]: Yep, trying to maintain the correct whitespace and indentation at least. But I need to indent more than just the original leading trivia.

File: src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs (hunk 3 above)
Comment (DavidBoike) [reply src/NServiceBus.Core.Analyzer/Sagas/RewriteConfigureHowToFindSagaFixer.cs:178]: Nope, vestigial from when I was fighting the whitespace dragon.
```suggestion
```

File: src/NServiceBus.Core.Analyzer.Tests/Sagas/RewriteConfigureHowToFindSagaFixerTests.cs (hunk 2 above)
Comment (DavidBoike) [reply src/NServiceBus.Core.Analyzer.Tests/Sagas/RewriteConfigureHowToFindSagaFixerTests.cs:301]: Nope
```suggestion
```

**Review by cquirosj — APPROVED** [review]
Can we have the branch rebased on the latest master before merging? So that when it gets merged the git log doesn't show ugly and hard-to-follow lines but instead just a nice "belly" line? :)

**Review by TravisNickels — APPROVED** [review]

**Pull request merged**